wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs

```diff
@@ -14,39 +18,65 @@ module Cardano.Wallet.WalletLayer.Types
+                                              Payment, Transaction, V1, Wallet, WalletAddress,
+                                              WalletId, WalletUpdate)
+import           Pos.Core (Address, TxAux)
 
 import           Cardano.Wallet.Kernel.Diffusion (WalletDiffusion (..))
+import           Cardano.Wallet.WalletLayer.Error (WalletLayerError (..))
 
 ------------------------------------------------------------
 -- Passive wallet layer
 ------------------------------------------------------------
 
+-- The response for the @WalletPassiveLayer@.
+type WalletLayerResponse m a = m (Either WalletLayerError a)
+
+-- A utility function for transforming between representations.
+monadThrowToEither :: forall m a. (MonadCatch m) => m a -> WalletLayerResponse m a
+monadThrowToEither = try
+
 -- | The passive wallet (data) layer. See @PassiveWallet@.
 data PassiveWalletLayer m = PassiveWalletLayer
     {
     -- * wallets
-      _pwlCreateWallet  :: NewWallet -> m Wallet
-    , _pwlGetWalletIds  :: m [WalletId]
-    , _pwlGetWallet     :: WalletId -> m (Maybe Wallet)
-    , _pwlUpdateWallet  :: WalletId -> WalletUpdate -> m Wallet
-    , _pwlDeleteWallet  :: WalletId -> m Bool
+      _pwlCreateWallet   :: NewWallet -> WalletLayerResponse m Wallet
+    , _pwlGetWalletIds   :: WalletLayerResponse m [WalletId]
+    , _pwlGetWallet      :: WalletId -> WalletLayerResponse m Wallet
+    , _pwlUpdateWallet   :: WalletId -> WalletUpdate -> WalletLayerResponse m Wallet
+    , _pwlDeleteWallet   :: WalletId -> WalletLayerResponse m ()
     -- * accounts
-    , _pwlCreateAccount :: WalletId -> NewAccount -> m Account
-    , _pwlGetAccounts   :: WalletId -> m [Account]
-    , _pwlGetAccount    :: WalletId -> AccountIndex -> m (Maybe Account)
-    , _pwlUpdateAccount :: WalletId -> AccountIndex -> AccountUpdate -> m Account
-    , _pwlDeleteAccount :: WalletId -> AccountIndex -> m Bool
+    , _pwlCreateAccount  :: WalletId -> NewAccount -> WalletLayerResponse m Account
+    , _pwlGetAccounts    :: WalletId -> WalletLayerResponse m [Account]
+    , _pwlGetAccount     :: WalletId -> AccountIndex -> WalletLayerResponse m Account
+    , _pwlUpdateAccount  :: WalletId -> AccountIndex -> AccountUpdate -> WalletLayerResponse m Account
+    , _pwlDeleteAccount  :: WalletId -> AccountIndex -> WalletLayerResponse m ()
     -- * addresses
-    , _pwlGetAddresses  :: WalletId -> m [Address]
+    , _pwlCreateAddress  :: NewAddress -> WalletLayerResponse m WalletAddress
+    , _pwlGetAddresses   :: WalletId -> Maybe AccountIndex -> WalletLayerResponse m [WalletAddress]
+    , _pwlIsAddressValid :: WalletAddress -> WalletLayerResponse m AddressValidity
+    -- * transactions
+    , _pwlCreateTx       :: (TxAux -> m Bool) -> Payment -> WalletLayerResponse m Transaction
+    , _pwlGetTxs         :: Maybe WalletId -> Maybe AccountIndex -> Maybe (V1 Address) -> WalletLayerResponse m [Transaction]
+    , _pwlEstimateFees   :: Payment -> WalletLayerResponse m EstimatedFees
```

Comment:
style note: putting the `::` on a newline will provide for cleaner diffs and make it easier to know what actually changed. eg:

```haskell
data X = X
    { foo
        :: Int -> IO ()
    , bar
        :: Double -> IO Char
    }
```